### PR TITLE
K8s/Playlist: Test dual write path

### PR DESCRIPTION
**Pull request by ryantxu**
This PR updates the dual write logic so that our integration tests work.

With this PR, we will be testing the storage engine in main 🎉

### CODE REVIEW

File: pkg/services/grafana-apiserver/rest/dualwriter.go [hunk 1]
```diff
@@ -67,10 +70,17 @@ func NewDualWriter(legacy LegacyStorage, storage Storage) *DualWriter {
...
 	if legacy, ok := d.legacy.(rest.Creater); ok {
-		_, err := legacy.Create(ctx, obj, createValidation, options)
+		created, err := legacy.Create(ctx, obj, createValidation, options)
 		if err != nil {
 			return nil, err
 		}
+		obj = created // write the updated version
+		rsp, err := d.Storage.Create(ctx, obj, createValidation, options)
+		if err != nil {
+			// TODO
```
Comment (ryantxu) [inline pkg/services/grafana-apiserver/rest/dualwriter.go:80]: @toddtreece -- thoughts on what to do here?  For now just a comment?

The next big step is a reconciliation pattern -- something that will make sure the legacy values are in sync with storage.  Maybe that is our fallback answer on how error cascading will work

File: pkg/services/grafana-apiserver/rest/dualwriter.go [hunk 2]
```diff
@@ -79,35 +89,92 @@ func (d *DualWriter) Create(ctx context.Context, obj runtime.Object, createValid
...
 	return d.Storage.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
 }
 
 // Delete overrides the default behavior of the Storage and delete from both the LegacyStorage and Storage.
 func (d *DualWriter) Delete(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
-	if legacy, ok := d.legacy.(rest.GracefulDeleter); ok {
-		_, _, err := legacy.Delete(ctx, name, deleteValidation, options)
-		if err != nil {
-			return nil, false, err
+	// Delete from storage *first* so the item is still exists if a failure happens
```
Comment (ryantxu) [inline pkg/services/grafana-apiserver/rest/dualwriter.go:143]: not tested in this PR, but I think we should flip the direction of possible errors.  keeping the value in SQL as long as possible

File: pkg/services/grafana-apiserver/rest/dualwriter.go [hunk 3]
```diff
@@ -54,23 +58,31 @@ type LegacyStorage interface {
...
 	if legacy, ok := d.legacy.(rest.Creater); ok {
-		_, err := legacy.Create(ctx, obj, createValidation, options)
+		created, err := legacy.Create(ctx, obj, createValidation, options)
 		if err != nil {
 			return nil, err
 		}
+		obj = created // write the updated version
+		rsp, err := d.Storage.Create(ctx, obj, createValidation, options)
+		if err != nil {
+			d.log.Error("unable to create object in duplicate storage", "error", err)
```
Comment (ryantxu) [inline pkg/services/grafana-apiserver/rest/dualwriter.go:83]: could use klog -- but we are not yet doing that elsewhere, so i kept an explicit logger here

**Comment by ryantxu** [comment]
@toddtreece -- ready for review here.

**Review by toddtreece — APPROVED** [review]

File: pkg/services/grafana-apiserver/storage/file/util.go [hunk 4]
```diff
@@ -55,6 +56,9 @@ func readDirRecursive(codec runtime.Codec, path string, newFunc func() runtime.O
 		return nil
 	})
 	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/grafana-apiserver/storage/file/util.go:59]: k8s has special handling for the not-found case vs general error, and this is how it knows the difference

[the remark this answers is not in the document]
Comment (toddtreece) [reply pkg/services/grafana-apiserver/storage/file/util.go:59]: ah ok, i think i remember simplifying some of the duplicate exist checking now, so maybe that's why these errors were happening

File: pkg/services/grafana-apiserver/rest/dualwriter.go [hunk 5]
```diff
@@ -54,23 +59,31 @@ type LegacyStorage interface {
...
 // Create overrides the default behavior of the Storage and writes to both the LegacyStorage and Storage.
 func (d *DualWriter) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
 	if legacy, ok := d.legacy.(rest.Creater); ok {
-		_, err := legacy.Create(ctx, obj, createValidation, options)
+		created, err := legacy.Create(ctx, obj, createValidation, options)
 		if err != nil {
 			return nil, err
 		}
+		obj = created // write the updated version
+		rsp, err := d.Storage.Create(ctx, obj, createValidation, options)
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/grafana-apiserver/rest/dualwriter.go:82]: Originally I thought we should *remove* it from legacy if either fails, but after more discussion this seems OK.

we could also just fall though to the flavor outside the clause, but I think we need at least the error handling here.

File: pkg/services/grafana-apiserver/rest/dualwriter.go [hunk 6]
```diff
@@ -79,35 +92,92 @@ func (d *DualWriter) Create(ctx context.Context, obj runtime.Object, createValid
...
 }
 
 // DeleteCollection overrides the default behavior of the Storage and delete from both the LegacyStorage and Storage.
 func (d *DualWriter) DeleteCollection(ctx context.Context, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions, listOptions *metainternalversion.ListOptions) (runtime.Object, error) {
-	if legacy, ok := d.legacy.(rest.CollectionDeleter); ok {
-		_, err := legacy.DeleteCollection(ctx, deleteValidation, options, listOptions)
-		if err != nil {
-			return nil, err
+	out, err := d.Storage.DeleteCollection(ctx, deleteValidation, options, listOptions)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/grafana-apiserver/rest/dualwriter.go:163]: isn't that what is happening (note the == nil 😄 )

[the remark this answers is not in the document]
Comment (toddtreece) [reply pkg/services/grafana-apiserver/rest/dualwriter.go:163]: by returning early, i mean the pattern i normally see is:

```go
if err != nil {
  return nil, err
}
```

not a huge deal, but if it's anything other than that, i spend extra time wondering what other logic is happening that required reversing the normal pattern

**Pull request merged**